### PR TITLE
Refactor code2seq

### PR DESCRIPTION
**Pull request by maximzubkov**

### CODE REVIEW

File: configs/__init__.py [hunk 1]
```diff
@@ -63,33 +56,67 @@ def get_code2seq_default_config(dataset_main_folder: str) -> Code2SeqConfig:
...
+        embedding_size=128, rnn_size=128, use_bi_rnn=True, embedding_dropout=0.25, rnn_num_layers=1, rnn_dropout=0.5
     )
-    decoder = DecoderConfig(
-        decoder_size=120, embedding_size=64, num_decoder_layers=1, rnn_dropout=0.5, teacher_forcing=1, beam_width=0
+    decoder_config = DecoderConfig(
+        decoder_size=320, embedding_size=128, num_decoder_layers=1, rnn_dropout=0.5, teacher_forcing=1, beam_width=0
     )
 
-    code2seq = Code2SeqConfig(
+    config = _get_default_hyperparams(dataset_main_folder)
```
[the remark this answers is not in the document]
Comment (SpirinEgor) [reply configs/__init__.py:69]: And change it in other functions.

File: model/__init__.py [hunk 2]
```diff
@@ -1 +1,3 @@
 from .code2seq import Code2Seq
+from .base_code_model import BaseCodeModel, StatisticType
```
Comment (SpirinEgor) [inline model/__init__.py:2]: Please, remove `StatisticType` we will work on it later

Comment (SpirinEgor) [reply model/__init__.py:2]: Also, I suggest to sort imports alphabetically.

File: model/base_code_model.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+from pytorch_lightning.core.lightning import LightningModule
+from torch.optim import Adam, Optimizer, SGD
+from torch.optim.lr_scheduler import ExponentialLR, _LRScheduler
+from torch.utils.data import DataLoader
+
+from configs import ModelHyperparameters
+from dataset import Vocabulary, create_dataloader, PathContextBatch
+from utils.metrics import SubtokenStatistic, ClassificationStatistic
+
+StatisticType = Union[SubtokenStatistic, ClassificationStatistic]
```
Comment (SpirinEgor) [inline model/base_code_model.py:15]: Still don't need this

File: model/base_code_model.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    def __init__(
+        self, config: ModelHyperparameters, vocab: Vocabulary, num_workers: int = 0,
+    ):
+        super().__init__()
+        self.save_hyperparameters()
+        self.config = config
+        self.vocab = vocab
+        self.num_workers = num_workers
+
+    def forward(
```
Comment (SpirinEgor) [inline model/base_code_model.py:28]: This is an abstract method, forward pass can be different for each model. Moreover, I don't sure, that code2class will use the same arguments, for example, it doesn't need target_sequnce for teacher forcing.

File: model/base_code_model.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        self,
+        samples: Dict[str, torch.Tensor],
+        paths_for_label: List[int],
+        output_length: int = None,
+        target_sequence: torch.Tensor = None,
+    ) -> torch.Tensor:
+        return self.decoder(self.encoder(samples), paths_for_label, output_length, target_sequence)
+
+    @abstractmethod
+    def _calculate_loss(self, logits: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
```
Comment (SpirinEgor) [inline model/base_code_model.py:38]: As I said in slack, this method used only in `code2seq`. For `code2class` you can call `CrossEntropyLoss` in one line, you don't need to move it in different method.

File: model/base_code_model.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        """Calculate loss"""
+        pass
+
+    @abstractmethod
+    def _compute_metrics(self, logits: torch.Tensor, labels: torch.Tensor) -> StatisticType:
+        pass
```
Comment (SpirinEgor) [inline model/base_code_model.py:44]: Let's implement without metrics

File: model/base_code_model.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    @abstractmethod
+    def _general_epoch_end(self, outputs: List[Dict], loss_key: str, group: str) -> Dict:
+        pass
+
+    @abstractmethod
+    def _get_progress_bar(self, log: Dict, group: str) -> Dict[str, float]:
+        pass
```
Comment (SpirinEgor) [inline model/base_code_model.py:52]: Also, it can be done in `training_step`

File: model/base_code_model.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+    def transfer_batch_to_device(self, batch: PathContextBatch, device: torch.device) -> PathContextBatch:
+        # Dict str -> torch.Tensor [seq length; batch size * n_context]
+        for k in batch.context:
+            batch.context[k] = batch.context[k].to(self.device)
+        # [seq length; batch size]
+        batch.labels = batch.labels.to(self.device)
+        return batch
+
+    def training_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
```
Comment (SpirinEgor) [inline model/base_code_model.py:121]: I suggest making this abstract.
1. Calculating logits may be different due to the different args (you don't need sequence length and ground truth labels for classification problem).
2. For classification problem, as I mentioned before, you don't need another method at all.
3. Different problems -- different metrics and logging to progress bar.

File: model/base_code_model.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        log = {"train/loss": loss}
+        with torch.no_grad():
+            statistic = self._compute_metrics(logits, batch.labels)
+
+        log.update(statistic.calculate_metrics(group="train"))
+        progress_bar = self._get_progress_bar(log, "train")
+
+        return {"loss": loss, "log": log, "progress_bar": progress_bar, "statistic": statistic}
+
+    def validation_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
```
Comment (SpirinEgor) [inline model/base_code_model.py:134]: This should be abstract too.

File: model/base_code_model.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        # [seq length; batch size; vocab size]
+        logits = self(batch.context, batch.contexts_per_label, batch.labels.shape[0])
+        loss = self._calculate_loss(logits, batch.labels)
+        with torch.no_grad():
+            statistic = self._compute_metrics(logits, batch.labels)
+
+        return {"val_loss": loss, "statistic": statistic}
+
+    def test_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
```
Comment (SpirinEgor) [inline model/base_code_model.py:143]: I'm not sure about this. It can stay here, but I think if we make it abstract it will be more accurate.

File: configs/__init__.py [hunk 11]
```diff
@@ -26,7 +28,7 @@ def get_preprocessing_config_code2class_params(dataset_name: str) -> Preprocessi
         dataset_name=dataset_name,
         max_path_length=9,
         max_name_parts=5,
-        max_target_parts=6,
+        max_target_parts=1,
```
[the remark this answers is not in the document]
Comment (maximzubkov) [reply configs/__init__.py:31]: But it was incorrectly defined previously because for classification problem we only need max_target_parts=1

[the remark this answers is not in the document]
Comment (maximzubkov) [reply configs/__init__.py:31]: But this change is done exactly in `get_preprocessing_config_code2class_params` which was incorrectly defined previously

File: model/base_code_model.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+from pytorch_lightning.core.lightning import LightningModule
+from torch.optim import Adam, Optimizer, SGD
+from torch.optim.lr_scheduler import ExponentialLR, _LRScheduler
+from torch.utils.data import DataLoader
+
+from configs import ModelHyperparameters
+from dataset import Vocabulary, create_dataloader, PathContextBatch
+
+
+class BaseCodeModel(LightningModule, metaclass=ABCMeta):
```
Comment (SpirinEgor) [inline model/base_code_model.py]: Let's use `ABC`:
>  Using ABC as a base class has essentially the same effect as specifying metaclass=abc.ABCMeta, but is simpler to type and easier to read

https://docs.python.org/3/whatsnew/3.4.html#abc

File: model/code2class.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        super().__init__(hyperparams, vocab, num_workers)
+        self.encoder = PathEncoder(
+            encoder_config,
+            decoder_config.classifier_input_size,
+            len(self.vocab.token_to_id),
+            self.vocab.token_to_id[PAD],
+            len(self.vocab.type_to_id),
+            self.vocab.type_to_id[PAD],
+        )
+        self.decoder = PathClassifier(decoder_config, len(self.vocab.label_to_id))
```
Comment (SpirinEgor) [inline model/code2class.py:32]: Maybe `classifier` will be better name

File: model/code2class.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
+from typing import Dict, List
+
+import torch
+import torch.nn.functional as F
+
+from configs import ModelHyperparameters, EncoderConfig, ClassifierConfig
+from dataset import Vocabulary, PathContextBatch
+from model.modules import PathEncoder, PathClassifier
+from utils.common import PAD
+from utils.metrics import ClassificationStatistic
```
Comment (SpirinEgor) [inline model/code2class.py:10]: And again, let's create classifier without any statistic

File: model/code2class.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+
+class Code2Class(BaseCodeModel):
+    def __init__(
+        self,
+        hyperparams: ModelHyperparameters,
+        vocab: Vocabulary,
+        num_workers: int,
+        encoder_config: EncoderConfig,
+        decoder_config: ClassifierConfig,
```
Comment (SpirinEgor) [inline model/code2class.py:21]: `classifier_config`

File: model/code2class.py [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        )
+        self.decoder = PathClassifier(decoder_config, len(self.vocab.label_to_id))
+
+    def _general_epoch_end(self, outputs: List[Dict], loss_key: str, group: str) -> Dict:
+        logs = {f"{group}/loss": torch.stack([out[loss_key] for out in outputs]).mean()}
+        logs.update(
+            ClassificationStatistic(len(self.vocab.label_to_id))
+            .union_statistics([out["statistic"] for out in outputs])
+            .calculate_metrics(group)
+        )
```
Comment (SpirinEgor) [inline model/code2class.py:43]: Remove this, please

File: model/code2class.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        progress_bar = {k: v for k, v in logs.items() if k in [f"{group}/loss", f"{group}/accuracy"]}
+
+        return {"val_loss": logs[f"{group}/loss"], "log": logs, "progress_bar": progress_bar}
+
+    def training_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
+        # [seq length; batch size; vocab size]
+        logits = self(batch.context, batch.contexts_per_label, batch.labels.shape[0], batch.labels)
```
Comment (SpirinEgor) [inline model/code2class.py]: Why do you need 3rd and 4th arguments for forward pass?

File: model/code2class.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+        loss = F.cross_entropy(logits, batch.labels.squeeze(0))
+        log = {"train/loss": loss}
+        with torch.no_grad():
+            statistic = ClassificationStatistic(len(self.vocab.label_to_id)).calculate_statistic(
```
Comment (SpirinEgor) [inline model/code2class.py:54]: Not now :)

File: model/base_code_model.py [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+        self, hyperparams: ModelHyperparameters, vocab: Vocabulary, num_workers: int = 0,
+    ):
+        super().__init__()
+        self.save_hyperparameters()
+        self.hyperparams = hyperparams
+        self.vocab = vocab
+        self.num_workers = num_workers
+
+    @abstractmethod
+    def forward(
```
Comment (SpirinEgor) [inline model/base_code_model.py]: Does it fit in one line?

File: model/code2seq.py [hunk 20]
```diff
@@ -1,68 +1,47 @@
...
-            decoder_config, len(vocab.label_to_id), vocab.label_to_id[SOS], vocab.label_to_id[PAD]
+            decoder_config, len(self.vocab.label_to_id), self.vocab.label_to_id[SOS], self.vocab.label_to_id[PAD]
         )
 
     def forward(
         self,
         samples: Dict[str, torch.Tensor],
         paths_for_label: List[int],
-        output_length: int,
+        output_length: int = None,
```
Comment (SpirinEgor) [inline model/code2seq.py:40]: It shouldn't be None

File: model/modules/path_classifier.py [hunk 21]
```diff
@@ -30,7 +30,13 @@ def __init__(self, config: ClassifierConfig, out_size: int):
         self.hidden_layers = nn.Sequential(*layers)
         self.classification_layer = nn.Linear(config.hidden_size, self.out_size)
 
-    def forward(self, encoded_paths: torch.Tensor, contexts_per_label: List[int],) -> torch.Tensor:
+    def forward(
+        self,
+        encoded_paths: torch.Tensor,
+        contexts_per_label: List[int],
+        output_length: int = None,
+        target_sequence: torch.Tensor = None,
```
Comment (SpirinEgor) [inline model/modules/path_classifier.py:38]: Do you add these arguments and don't use it?)

File: train.py [hunk 22]
```diff
@@ -66,6 +83,7 @@ def train(dataset_name: str, num_workers: int = 0, is_test: bool = False, resume
     arg_parser.add_argument("--n_workers", type=int, default=0)
     arg_parser.add_argument("--test", action="store_true")
     arg_parser.add_argument("--resume", type=str, default=None)
+    arg_parser.add_argument("--model", type=str, default="code2seq")
```
Comment (SpirinEgor) [inline train.py:86]: Let's set it as the first positional argument. Also, use choice to specify possible values (https://docs.python.org/3/library/argparse.html#choices)

File: utils/metrics.py [hunk 23]
```diff
@@ -1,5 +1,6 @@
 from dataclasses import dataclass
 from typing import Dict, List
+from .cls_metrics import ClassificationStatistic
```
Comment (SpirinEgor) [inline utils/metrics.py:3]: Please, remove all `ClassificationStatistic` from this pr

File: model/base_code_model.py [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+        self, hyperparams: ModelHyperparameters, vocab: Vocabulary, num_workers: int = 0,
+    ):
+        super().__init__()
+        self.save_hyperparameters()
+        self.hyperparams = hyperparams
+        self.vocab = vocab
+        self.num_workers = num_workers
+
+    @abstractmethod
+    def forward(self, samples: Dict[str, torch.Tensor], paths_for_label: List[int],) -> torch.Tensor:
```
Comment (SpirinEgor) [inline model/base_code_model.py:26]: Do we need to add `*args` here? For example, `code2seq` requires more args.

File: train.py [hunk 25]
```diff
@@ -8,44 +8,61 @@
...
-    config = config_function(dataset_main_folder)
-
-    model = Code2Seq(config, vocab, num_workers)
+    if model_name == "code2seq":
+        config_function = get_code2seq_test_config if is_test else get_code2seq_default_config
+        hyperparams, encoder_config, decoder_config = config_function(dataset_main_folder)
+        model = Code2Seq(hyperparams, vocab, num_workers, encoder_config=encoder_config, decoder_config=decoder_config)
+    elif model_name == "code2class":
+        config_function = get_code2class_test_config if is_test else get_code2class_default_config
+        hyperparams, encoder_config, decoder_config = config_function(dataset_main_folder)
```
Comment (SpirinEgor) [inline train.py:37]: `classifier_config`

File: train.py [hunk 26]
```diff
@@ -8,44 +8,61 @@
...
+    else:
+        raise ValueError(f"Model {model_name} is not supported")
 
     # define logger
     wandb_logger = WandbLogger(project=f"code2seq-{dataset_name}", log_model=True, offline=is_test)
     wandb_logger.watch(model)
-    wandb_logger.log_hyperparams(asdict(config))
+    wandb_logger.log_hyperparams(asdict(hyperparams))
+    wandb_logger.log_hyperparams(asdict(encoder_config))
+    wandb_logger.log_hyperparams(asdict(decoder_config))
```
Comment (SpirinEgor) [inline train.py:49]: You can remove this because we have `self.save_hyperparameters()` in each model. At the moment, parameters are logged twice

File: model/code2class.py [hunk 27]
```diff
@@ -15,6 +15,7 @@ def __init__(
         self, config: Code2ClassConfig, vocab: Vocabulary, num_workers: int,
     ):
         super().__init__(config.hyperparams, vocab, num_workers)
+        self.save_hyperparameters()
```
Comment (SpirinEgor) [inline model/code2class.py:18]: Doesn't that log hyperparameters twice? Here and in base class
